### PR TITLE
fix: support passing collection in json_body

### LINKED ISSUE #1746: Java SDK: ill-deserialization of RequestBody typed to List<T>
This was initially reported in the Kalix Slack channel: https://kalixworld.slack.com/archives/C05CBK0VCKH/p1689858932842709

Given a method defined as

```java
@PostMapping("/createMore")
public Effect<String> createMore(@RequestBody List<Customer> customers) {...}
```

When we call this endpoint and pass a json array:

```json
[
    {
        "customerId": "one",
        "email": "test@example.com",
        "name": "Test Testsson",
        "address": {
            "street": "Teststreet 25",
            "city": "Testcity"
        }
    },
    {
        "customerId": "two",
        "email": "test2@example.com",
        "name": "Test Testsson",
        "address": {
            "street": "Teststreet 25",
            "city": "Testcity"
        }
    }
]
```

The Kalix proxy will pass the payload as is to the user application. That's correct. 

On the SDK side, we receive the payload and we need to dispatch to method `createMore`, but at this point, we identify the `@RequestBody` as being of type `List` (not parameterized) and we have no type information about `Customer`. As a consequence, Jackson will deserialize the list into a List of LinkedHashMap (`List<LinkedHashMap>`). 

CleanShot 2023-07-24 at 11 28 31@2x

Later on, when the list is passed to the user function, it's casted to `List<Customer>`, but we don't have `Customer` inside it. And that will cause a ClassCastException as soon as we try to access any element in the list.

### PR DESCRIPTION
**Pull request by octonato**
Fixes #1746

### CODE REVIEW

File: sdk/java-sdk-spring/src/it/java/com/example/wiring/SpringSdkIntegrationTest.java [hunk 1]
```diff
@@ -167,36 +167,62 @@ public void verifyEchoActionWiring() {
...
+    var message = List.of(new Message("foo"), new Message("bar"));
+    Message response = execute(
+      componentClient.forAction()
+        .call(EchoAction::stringMessageConcatRequestBodyWithSeparator)
+        .params("/", message)
+    );
+
+    assertThat(response.text).isEqualTo("foo/bar");
+  }
+
```
Comment (octonato) [inline sdk/java-sdk-spring/src/it/java/com/example/wiring/SpringSdkIntegrationTest.java:218]: There are new tests. The other changes in this file are only formatting.

File: sdk/java-sdk-spring/src/it/java/com/example/wiring/actions/echo/BodyWithPrimitives.java
Comment (octonato) [inline sdk/java-sdk-spring/src/it/java/com/example/wiring/actions/echo/BodyWithPrimitives.java:1]: was not being used

**Review by efgpinto — APPROVED** [review]
LGTM

File: sdk/java-sdk-spring/src/main/scala/kalix/javasdk/impl/ComponentDescriptor.scala [hunk 2]
```diff
@@ -501,6 +517,16 @@ private[kalix] object ComponentDescriptor {
     }
   }
 
+  private def collectionBodyFieldExtractors[T](paramType: ParameterizedType): ExtractorCreator =
+    new ExtractorCreator {
+      override def apply(descriptor: Descriptors.Descriptor): ParameterExtractor[DynamicMessageContext, AnyRef] = {
+        // since we only support collections, the is only one type param at idx 0
+        val cls = paramType.getActualTypeArguments()(0).asInstanceOf[Class[T]]
+        val collectionClass = paramType.getRawType.asInstanceOf[Class[java.util.Collection[T]]]
+        new ParameterExtractors.CollectionBodyExtractor(descriptor.findFieldByNumber(1), cls, collectionClass)
```
[the remark this answers is not in the document]
Comment (octonato) [reply sdk/java-sdk-spring/src/main/scala/kalix/javasdk/impl/ComponentDescriptor.scala:526]:
```suggestion
        // json_body field is always on position 1 in the synthetic request
        new ParameterExtractors.CollectionBodyExtractor(descriptor.findFieldByNumber(1), cls, collectionClass)
```

When we get a Proto message with a request body, the `json_body` is always the first field. We did that on purpose to make it easier to pick it. 

Note that 1 here is not an array index, but Protobuf field index (that starts with 1).

[the remark this answers is not in the document]
Comment (octonato) [reply sdk/java-sdk-spring/src/main/scala/kalix/javasdk/impl/ComponentDescriptor.scala:526]: Adding a comment to make it more clear.

**Pull request merged**